metricbeat/module/windows/wmi/wmi.go

```diff
@@ -0,0 +1,212 @@
+		}
+	}
+	return false
+}
+
+// Fetch method implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+
+	sm := wmi.NewWmiSessionManager()
+	defer sm.Dispose()
+
+	// To optimize performance and reduce overhead, we create a single session
+	// for each unique WMI namespace. This minimizes the number of session creations
+	for namespace, queries := range m.config.NamespaceQueryIndex {
+
+		session, err := sm.GetSession(namespace, m.config.Host, "", m.config.User, m.config.Password)
+
+		if err != nil {
+			return fmt.Errorf("could not initialize session %w", err)
+		}
+		_, err = session.Connect()
+		if err != nil {
+			return fmt.Errorf("could not connect session %w", err)
+		}
+		defer session.Dispose()
+
+		// We create a shared conversion table for entries with the same schema.
+		// This avoids repeatedly fetching the schema for each individual entry, improving efficiency.
+		conversionTable := make(map[string]WmiStringConversionFunction)
+
+		for _, queryConfig := range queries {
+
+			query := queryConfig.QueryStr
+
+			rows, err := ExecuteGuardedQueryInstances(session, query, m.config.WarningThreshold, m.Logger())
+
+			if err != nil {
+				m.Logger().Warn("Could not execute query: %v", err)
+				continue
+			}
+
+			defer wmi.CloseAllInstances(rows)
+
+			for _, instance := range rows {
+				event := mb.Event{
+					MetricSetFields: mapstr.M{
+						"class":     queryConfig.Class,
+						"namespace": m.config.Namespace,
+						"host":      m.config.Host,
+					},
+				}
+
+				if m.config.IncludeQueries {
+					event.MetricSetFields.Put("query", query)
+				}
+
+				// Get only the required properties
+				properties := queryConfig.Fields
+
+				// If the Fields array is empty we retrieve all fields
+				if len(queryConfig.Fields) == 0 {
+					properties = instance.GetClass().GetPropertiesNames()
+				}
+
+				for _, fieldName := range properties {
+					fieldValue, err := instance.GetProperty(fieldName)
+					if err != nil {
+						m.Logger().Error("Unable to get propery by name: %v", err)
+						continue
+					}
+
+					if m.shouldSkipNilOrEmptyValue(fieldValue) {
+						continue
+					}
+
+					// The default case, we simply return what we got
+					finalValue := fieldValue
+
+					// The script API of WMI returns strings for uint64, sint64, datetime
+					// Link: https://learn.microsoft.com/en-us/windows/win32/wmisdk/querying-wmi
+					// As a user, I want to have the right CIM_TYPE in the final document
+					//
+					// Example: in the query: SELECT * FROM Win32_OperatingSystem
+					// FreePhysicalMemory is a string, but it should be an uint64
+					if RequiresExtraConversion(fieldValue) {
+						convertFun, ok := conversionTable[fieldName]
+						// If the function is not found let us fetch it and cache it
```

Comment:
what's the reason we do this, instead of just calling the functions?

Reply:
> If you are referring to the if:

I am doing this because there are many cases where we don't apply a transformation (example all integers in all architectures but 64).

> If you are referring to the caching of the function
There are queries with more than one line as a result (but share the same schema) and without the caching we would have the "CIMType" workaround for every string.

Reply:
it only saves cycles if we have fields with the same names in the same collection though, right? are the field names always unique between namespaces? because this would apply the same conversion there too i think (same field name, different namespace).

Reply:
Wow, great catch! This is (surely) unique for query but not for Namespace (or at least I would not do this assumption).

Let me move the definition at the query level. This will save cycles if we have X results for the same query.

Reply:
Addressed in https://github.com/elastic/beats/pull/42017/commits/b01e730e5b97700de1fda209c65816853bcec6dd